lib/stages/ReadGain.cpp

```diff
@@ -199,34 +195,41 @@ void ReadGain::read_gain_tracking() {
-        ptr_myfile = fopen(filename, "rb");
-        if (ptr_myfile == nullptr) {
-            WARN("GPU Cannot open gain file {:s}", filename);
-            all_beams_successful_update = false;
-            for (uint i = 0; i < _num_elements; i++) {
-                out_frame_tracking[(b * _num_elements + i) * 2] = default_gains[0];
-                out_frame_tracking[(b * _num_elements + i) * 2 + 1] = default_gains[1];
-            }
-        } else {
-            if (_num_elements
-                != fread(&out_frame_tracking[b * _num_elements * 2], sizeof(float) * 2,
-                         _num_elements, ptr_myfile)) {
-                WARN("Gain file ({:s}) wasn't long enough! Something went wrong, using default "
-                     "gains",
-                     filename);
+    {
+        std::lock_guard<std::mutex> lock(mux);
+        while (_gain_dir_tracking.size() > 0) {
+            std::pair<uint8_t, std::string> beam = _gain_dir_tracking.front();
+            _gain_dir_tracking.pop();
+            uint8_t beam_id = beam.first;
+            snprintf(filename, sizeof(filename), "%s/quick_gains_%04d_reordered.bin",
+                     beam.second.c_str(), freq_idx);
+            INFO("Tracking Beamformer Loading gains from {:s}", filename);
+            ptr_myfile = fopen(filename, "rb");
+            if (ptr_myfile == nullptr) {
+                WARN("GPU Cannot open gain file {:s}", filename);
                 all_beams_successful_update = false;
                 for (uint i = 0; i < _num_elements; i++) {
-                    out_frame_tracking[(b * _num_elements + i) * 2] = default_gains[0];
-                    out_frame_tracking[(b * _num_elements + i) * 2 + 1] = default_gains[1];
+                    out_frame_tracking[(beam_id * _num_elements + i) * 2] = default_gains[0];
+                    out_frame_tracking[(beam_id * _num_elements + i) * 2 + 1] = default_gains[1];
+                }
+            } else {
+                if (_num_elements
+                    != fread(&out_frame_tracking[beam_id * _num_elements * 2], sizeof(float) * 2,
+                             _num_elements, ptr_myfile)) {
+                    WARN("Gain file ({:s}) wasn't long enough! Something went wrong, using default "
+                         "gains",
+                         filename);
+                    all_beams_successful_update = false;
+                    for (uint i = 0; i < _num_elements; i++) {
+                        out_frame_tracking[(beam_id * _num_elements + i) * 2] = default_gains[0];
+                        out_frame_tracking[(beam_id * _num_elements + i) * 2 + 1] =
+                            default_gains[1];
+                    }
                 }
+                fclose(ptr_myfile);
             }
-            fclose(ptr_myfile);
-        }
-    } // end beam
+        } // end beam
+    }
     if (all_beams_successful_update) {
         gains_last_update_success_metric.labels({"tracking"}).set(1);
```

Comment:
This metric should be per-beam now.  Otherwise a successful beam update on say beam 1, will overwrite a failure from say beam 0 before.   Alternatively it needs to track the quality of each been individually and then report this overall statistic.   However given that we are moving to each project having control over each beam, I'm not sure how useful the global statistic is.